rfcs/0097-no-read-store-dir.md

```diff
@@ -0,0 +1,130 @@
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Losing the read (r) bit means that you can't list the files inside the store.
+The execute (x) bit allows us to `cd` to it and also access paths inside the store.
+
+E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory
+inside the store, and not the store itself, but an unprivileged user can't `ls /nix/store` to find the system configuration.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It might be a slight annoyance since shell completion won't work in the `/nix/store` anymore, e.g.
+if you have some hash 48914, you wouldn't be able to type `/nix/store/48914<tab>` to get the full path anymore.
+This is likely the biggest drawback.
+
+External tooling that does a traversal of the nix store (`find`, `du -s`, `ncdu`) would need `sudo` or explicitly given permissions.
+
+A sandboxed program could still have some idea of how the host machine is used by checking each store path that Hydra has ever built,
+then it could estimate whether the host machine has e.g. Tor on it.
+
+Attempting to have security comparable to bind-mounting each necessary store path inside a sandbox is not easy,
+since you need to be careful not to leak paths to the sandbox through global state, for example `/proc/cmdline`.
+There is no easy way to know whether some path is readable from inside the sandbox, since the problem is
+essentially proving that the sandbox doesn't have access to some secret that in this case is the store path itself.
+Since Unix software has not traditionally treated paths as secrets, this is a bit problematic if you're
+trying to make sure your sandbox can't access some path. We also can not know how future kernel interfaces and such
+will be, and they will likely not have the same considerations for "paths as secrets" as us.
+Essentially, you have to consider the store paths as encodings of their contents wrt. sandboxes.
+
+The above points about sandboxed programs are not negative effects of changing the permissions of /nix/store and such,
+but rather considerations for making use of this change in the context of sandboxes.
+
+# Alternatives
+[alternatives]: #alternatives
+
+If Nix was made to not reset the permissions of /nix/store back to 1775, users who want this change could
+do it themselves by simply putting this into their configuration.nix:
+```nix
+{
+  system.activationScripts.chmod-store.text = ''
+    ${pkgs.util-linux}/bin/unshare -m ${pkgs.bash}/bin/sh -c '${pkgs.util-linux}/bin/mount -o remount,rw /nix/store ; ${pkgs.coreutils}/bin/chmod 1771 /nix/store ; ...'
+  '';
+}
+```
+
+Traditional sandboxing could be seen as an alternative to this, however, they are not mutually exclusive,
+since this change is fundamentally just changing some permissions in NixOS.
+Improving the permissions for /nix/store will not decrease security for traditional sandboxes.
+
+# Unresolved questions
```

Comment:
How does this affect other use cases of Nix, like on non-NixOS or single-user installations?

Reply:
It won't be any different. Of course, on non-NixOS systems the permissions of /proc/cmdline, etc. are the responsibility of that distro, but everything else should be exactly the same.